File: Difficulty: Medium/Equalize the Towers/equalize-the-towers.cpp
class Solution {
  public:
    int findCost(int h, vector<int>& heights, vector<int>& cost) {
        int ans = 0;
        
        for(int i=0; i<heights.size(); i++) {
            ans += cost[i] * abs(heights[i] - h);
        }
        
        return ans;
    }
    
    int minCost(vector<int>& heights, vector<int>& cost) {
        int minH = INT_MAX, maxH = INT_MIN;
        int ans;
        
        for(int h : heights) {
            minH = min(minH, h);
            maxH = max(maxH, h);
        }
        
        while(minH <= maxH) {
            int midH = minH + (maxH - minH) / 2;
            
            int leftCost = findCost(midH - 1, heights, cost);
            int midCost = findCost(midH, heights, cost);
            int rightCost = findCost(midH + 1, heights, cost);
            
            if(midCost <= leftCost && midCost <= rightCost) {
                ans = midCost;
                break;
            } else if(rightCost <= midCost && midCost <= leftCost) {
                minH = midH + 1;
            } else if(leftCost <= rightCost && leftCost <= midCost) {
                maxH = midH - 1;
            }
        }
        
        return ans;
    }
};

File: Difficulty: Medium/Equalize the Towers/equalize-the-towers.java
class Solution {
    public int calculateCost(int[] heights, int[] costs, int h) {
        int cost = 0;
        for(int i=0; i<heights.length; i++) {
            cost += Math.abs(heights[i] - h) * costs[i];
        }
        return cost;
    }
    
    public int minCost(int[] heights, int[] cost) {
        int low = Arrays.stream(heights).min().getAsInt();
        int high = Arrays.stream(heights).max().getAsInt();
        
        while(low <= high) {
            int mid = (low + high) >>> 1;
            int c0 = calculateCost(heights, cost, mid - 1);
            int c1 = calculateCost(heights, cost, mid);
            int c2 = calculateCost(heights, cost, mid + 1);
            
            if(c1 <= c0 && c1 <= c2) return c1;
            else if(c0 >= c1 && c1 >= c2) low = mid + 1;
            else high = mid - 1;
        }
        
        return -1;
    }
}
